Upgrade to lucene-9.0.0-snapshot-32a0a16aff0

romseygeek opened the pull request:
This commit upgrades elasticsearch 8.0 to use a snapshot of
Lucene 9.0.0.

romseygeek committed:
Upgrade to lucene-9.0.0-snapshot-a12260eb950

elasticmachine commented:
Pinging @elastic/es-analytics-geo (Team:Analytics)

elasticmachine commented:
Pinging @elastic/es-search (Team:Search)

elasticmachine commented:
Pinging @elastic/es-distributed (Team:Distributed)

romseygeek commented:
There are a few nocommits, but surprisingly few.  Everything compiles.

Still to do:
- I have silently remapped the `SIMPLEFS` store type to just use NIOFS.  I'm not sure if this is correct and we will just deprecate the option, or if it should throw an error.
- The Version 7.14 constant has the wrong lucene version on it, because we haven't released 8.9 yet
- There are a bunch of classes around segment memory accounting that I have left in but are now essentially redundant, eg RamAccountingRefreshListener and SegmentStats.  Segment itself also now has unused memoryInBytes and ramTree fields, which I am guessing should be removed but will need some careful BWC handling.
- CuckooFilter uses PackedInts.getReader() and save(), both of which are now gone.  Again this was being used as part of the Streamable impl and so will need BWC to be addressed when this is reimplemented.

dnhatn commented:
> I have silently remapped the SIMPLEFS store type to just use NIOFS. I'm not sure if this is correct and we will just deprecate the option, or if it should throw an error.

I think we should deprecate the `simplefs` store option in 7.x and remove it in 8.x.

iverase commented:
Hey @romseygeek,

I have been looking into the CuckooFilter issue. I think we should have our or version of PackedInts.Mutable, but before anything we need to improve our bwc testing for that part of the code. It seems it is never exercise so we can easily introduce a bug: https://github.com/elastic/elasticsearch/pull/73585

romseygeek committed:
Merge remote-tracking branch 'origin/master' into lucene-9-upgrade

romseygeek committed:
upgrade to new snapshot; precommit happy

jpountz committed:
Merge branch 'master' into lucene-9-upgrade

jpountz committed:
More compile fixes.

jpountz committed:
Undo unnecessary change.

dnhatn committed:
Merge remote-tracking branch 'elastic/master' into lucene-9-upgrade

dnhatn committed:
fix compile

dnhatn commented:
@romseygeek We need to build a new Lucene 9.0 snapshot as the current snapshot doesn't have the version 8.9.0, which is being used in 7.14 and 7.15.

romseygeek committed:
Merge remote-tracking branch 'origin/master' into lucene-9-upgrade

romseygeek committed:
compilation + precommit

romseygeek committed:
small updates

romseygeek committed:
Merge remote-tracking branch 'romseygeek/lucene-9-upgrade' into lucene-9-upgrade

romseygeek committed:
iter

romseygeek committed:
Merge remote-tracking branch 'origin/master' into lucene-9-upgrade

romseygeek committed:
another upgrade...

romseygeek committed:
Lucene test naming conventions

romseygeek committed:
schema consistency

romseygeek committed:
Translog + a few other places need to use CodecUtil.read/writeBEInt for BWC

romseygeek committed:
TopDocs.merge doesn't set shardIndex; new filters; codecs; math is hard; spi

romseygeek committed:
iter

romseygeek committed:
more precommit

romseygeek committed:
Merge remote-tracking branch 'origin/master' into lucene-9-upgrade

romseygeek committed:
move and rename org.apache.lucene tests

romseygeek committed:
precommit

romseygeek committed:
security permission for SPIClassIterator

romseygeek committed:
analysis fixes

romseygeek committed:
some more BEInt/Long changes

romseygeek committed:
don't need to set shard index in merge

romseygeek committed:
iter

romseygeek commented:
This pretty close, but there are some things I still need to fix - more eyes welcome

- sorting by a numeric field now results in a call to Points due to LUCENE-9280.  This means that the responses in `60_field_usage` are different across lucene versions - I guess the simple solution is just to add this test to the rest backwards compat exclusions list
- the docs integ tests are failing with a java security access exception, but I am struggling to see why given that the exact same code seems to work in all other integ tests
- I've changed a bunch of stuff to use `CodecUtil.readBEInt()` to work around the change in endian-ness for DataInput and DataOutput but Ignacio has pointed me towards `EndiannessReverserUtil` which is almost certainly a better option here.

iverase commented:
@romseygeek I open https://github.com/elastic/elasticsearch/pull/76162 to replace the usages of DataInput / DataOutput with StreamInput / StreamOutput in geoshape doc values. WE can do the same later with histogram doc values.

romseygeek commented:
@iverase that looks great, thank you! I wonder if we can use the same classes for the translog to avoid EndiannessReverseUtil there as well?

iverase commented:
I think we need to use EndiannessReverseUtil in cases where we are writing files with headers and footers. A good example is the checkpoint class. I think we need to increase the version on the header so when reading a file we can check the version and if it is an old version then we wrap the DataInput?

romseygeek committed:
Merge remote-tracking branch 'origin/master' into lucene-9-upgrade

romseygeek committed:
Use EndiannessReverserUtil where necessary

romseygeek committed:
Don't run field usage API tests on pre-8 versions due to changes in points usage

romseygeek committed:
Merge remote-tracking branch 'origin/master' into lucene-9-upgrade

romseygeek committed:
precommit

romseygeek committed:
Merge remote-tracking branch 'origin/master' into lucene-9-upgrade

romseygeek committed:
Merge remote-tracking branch 'origin/master' into lucene-9-upgrade

romseygeek committed:
correct version; collapse tests to use STRING rather than STRING_VAL

romseygeek committed:
comment out check for points usage

romseygeek committed:
Merge remote-tracking branch 'origin/master' into lucene-9-upgrade

romseygeek committed:
rest compat for field usage stats

jpountz commented on server/src/main/java/org/elasticsearch/index/fielddata/plain/PagedBytesIndexFieldData.java:
```diff
@@ -9,8 +9,8 @@
-import org.apache.lucene.codecs.blocktree.FieldReader;
-import org.apache.lucene.codecs.blocktree.Stats;
+import org.apache.lucene.codecs.lucene90.blocktree.FieldReader;
+import org.apache.lucene.codecs.lucene90.blocktree.Stats;
```
Now that blocktree is versioned and that fielddata isn't used much anymore, I'm tempted to remove the optimization that uses these two classes. Otherwise we're going to have a testing gap with pre-Lucene 9 indices, for which this optimization doesn't work, and that our tests don't exercise much.

romseygeek replied:
Removed in 01b3050c02a09b5d684277ff49c2684a0c5fdf97

jpountz commented on server/src/main/java/org/elasticsearch/index/store/Store.java:
```diff
@@ -1295,6 +1303,19 @@ public void seek(long pos) throws IOException {
+                readBytes(skipBuffer, 0, step);
+                skipped += step;
+            }
+        }
```
It looks buggy to me that this implementation is not taking into account `verifiedPosition`?

jpountz replied:
I think we might be able to do something simpler here by leaving `skipBytes` unimplemented (so that it uses the implementation from the parent class, which calls seek) and changing `seek` to use `super.seek` rather than `skipBytes` to skip over some bytes.

romseygeek replied:
++, I've done this in 126f0c05b00e787665fbb6669903f0509eb44a37

romseygeek committed:
access control on analysis plugin reloading

romseygeek committed:
Merge remote-tracking branch 'romseygeek/lucene-9-upgrade' into lucene-9-upgrade

romseygeek committed:
Merge remote-tracking branch 'origin/master' into lucene-9-upgrade

romseygeek committed:
upgrade ICU from 62.1 to 68.2

romseygeek committed:
increment version on checkpoint and use endiannessreverserutil on old versions

romseygeek committed:
Remove FieldReader breaker optimization that only works for lucene 9 indices

romseygeek committed:
better verifying checksum impl

romseygeek commented:
> I think we need to use EndiannessReverseUtil in cases where we are writing files with headers and footers. A good example is the checkpoint class. I think in this case we need to increase the version on the header so when reading a file we can check the version and if it is an old version then we wrap the DataInput?

I've done this in e66ec36e18e31d9437de8e605c20ee6c64f1538b

romseygeek committed:
Merge remote-tracking branch 'origin/master' into lucene-9-upgrade

iverase commented:
> I've done this in e66ec36

Thanks @romseygeek. I think we need to look into other files that are reading / writing files. For example:

KeyStoreWrapper
MetadataStateFormat
ChecksumBlobStoreFormat
Store
TranslogHeader

romseygeek committed:
update completion postings format to 90

romseygeek committed:
Bump metadata versions where possible to avoid EndiannessReverserUtil

romseygeek commented:
I've bumped the versions on KeyStoreWrapper and MetadataStateFormat.  ChecksumBlobStoreFormat and TranslogHeader are already using StreamInput/StreamOutput so there's no need to change things there, and the corruption marker in Store only writes vInts, which haven't changed format.

romseygeek commented:
Actually KeyStoreWrapper is more complicated, because in some places it uses ChecksumIndexInput and in other places it uses java's build in DataInput.  I think I'll revert this last commit for now and we can address it in a followup?

romseygeek committed:
Revert "Bump metadata versions where possible to avoid EndiannessReverserUtil"

This reverts commit 77fc68e1ced6f8fe725fe1074b43fd7a0a5e324d.

romseygeek commented:
These last failures appear to be due to a problem with sorting on `short` fields.  We're picking up a LongNumericComparator somewhere and this is confusing the BKDReader which is only expecting to return 2 bytes of data at a time.

jpountz commented:
I wonder if it's due to the fact that we pass `SortField.Type.LONG` to all integer fields, which expects fields to be indexed with `LongPoint`, while `byte`, `short` and `integer` fields are actually indexed with `IntPoint`. Should these fields move to `SortField.Type.INT` in `IndexNumericFieldData.NumericType`? cc @mayya-sharipova @jimczi

romseygeek commented:
> Should these fields move to SortField.Type.INT in IndexNumericFieldData.NumericType? 

This will solve things, but we also need to add an IntValuesComparatorSource.  I think I might pull this out into a separate smaller change to make it more easily reviewable.

jpountz commented:
I merged a change that should make this sort of issue more easily detected https://github.com/apache/lucene/pull/283.

> pull this out into a separate smaller change

+1

romseygeek committed:
Merge remote-tracking branch 'origin/master' into lucene-9-upgrade

romseygeek committed:
Further upgrades

romseygeek committed:
new version

romseygeek committed:
Merge remote-tracking branch 'origin/master' into lucene-9-upgrade

romseygeek committed:
don't run field usage API in bwc mode, output values are different between lucene versions

romseygeek committed:
haversin in meters, not kilometers

romseygeek committed:
change in max clause count behaviour

romseygeek committed:
Merge remote-tracking branch 'origin/master' into lucene-9-upgrade

romseygeek committed:
composite agg tests

romseygeek committed:
Merge remote-tracking branch 'origin/master' into lucene-9-upgrade

romseygeek committed:
Merge remote-tracking branch 'origin/master' into lucene-9-upgrade

romseygeek committed:
Revert "haversin in meters, not kilometers"

This reverts commit f2fef47488b51e04c89c06fb1a881a40dadefe0d.

romseygeek committed:
pickup sort opto fix

romseygeek committed:
Allow async check index in searchable snapshot tests

dnhatn commented:
@elasticmachine test this please

dnhatn committed:
Merge branch 'master' into lucene-9-upgrade

dnhatn committed:
Disable sort optimization while query Lucene changes

dnhatn committed:
update shas

dnhatn committed:
fix tests

dnhatn committed:
fix javadocs

dnhatn committed:
awaitfix testRecovery

dnhatn commented:
@romseygeek CI is green now 🎉, but we still have two issues:

1. LuceneChangesSnapshot doesn't play well with the sort optimization. I have disabled it in https://github.com/romseygeek/elasticsearch/commit/a6aee034ceb3e93c30db3c17ac3b39d711f1d728, and it should be resolved by https://github.com/elastic/elasticsearch/pull/77907.

2. I have to use Lucene 8.9 for 7.x as we don't have version 8.10 in this snapshot. Can you please add version 8.10 and build a new snapshot? We will remove some awaitfixs after that. Thank you!

ChrisHegarty commented:
This PR contains some changes to server/src/main/java/org/apache/lucene/search/XCombinedFieldQuery.java, but can that class simply be deleted? Since LUCENE-9999 has been fixed in Lucene v8.10.  Alternatively, its removal can be assessed independently after this PR is merged.

dnhatn commented:
@ChrisHegarty Good point. I will open a separate PR to remove that class.

dnhatn commented:
@ChrisHegarty I've opened https://github.com/elastic/elasticsearch/pull/77978 to remove the XCombinedFieldQuery workaround.

dnhatn committed:
Merge branch 'master' into lucene-9-upgrade

dnhatn committed:
fix compile

romseygeek committed:
Merge remote-tracking branch 'origin/master' into lucene-9-upgrade

romseygeek committed:
New snapshot, including lucene 8.10 version constant

romseygeek committed:
precommit

romseygeek commented:
@dnhatn awesome, thank you! I have updated the snapshot to latest lucene main and added a version 8.10 constant.  Fingers crossed that CI is still happy...

romseygeek committed:
Merge remote-tracking branch 'origin/master' into lucene-9-upgrade

romseygeek committed:
japanese completion filter

iverase commented on server/src/main/java/org/elasticsearch/action/search/TransportSearchHelper.java:
```diff
@@ -30,13 +31,15 @@ static InternalScrollSearchRequest internalScrollSearchRequest(ShardSearchContex
-        try (RAMOutputStream out = new RAMOutputStream()) {
+        try {
+            // TODO replace with BytesStreamOutput to avoid endiannessreverser
+            ByteBuffersDataOutput out = new ByteBuffersDataOutput();
```
Are we leaving this as a follow up?

romseygeek replied:
No, good catch, I had intended to update this one.  Done in 46fc8c1

iverase reviewed (approved):
LGTM. I have only reviewed the bits affected for the change on endianness of Lucene directory API. I think we should revisit the strategy on KeyStoreWrapper and MetadataStateFormat but it can be done as a follow up.

romseygeek committed:
No need for endianness shenanigans in TransportSearchHelper

dnhatn committed:
Merge branch 'master' into lucene-9-upgrade

dnhatn committed:
Remove NORELEASE

dnhatn committed:
Do not remove max_merge_at_once_explicit

jtibshirani commented on server/src/main/java/org/elasticsearch/index/engine/RecoverySourcePruneMergePolicy.java:
```diff
@@ -258,6 +247,48 @@ public StoredFieldsReader clone() {
 
         }
 
+        private static class FilterStoredFieldVisitor extends StoredFieldVisitor {
```
Do we need this new private class when we already have one in `org.elasticsearch.search.internal`?

dnhatn committed:
use existing filtered store fields visitor

jtibshirani reviewed (approved):
I looked through with a focus on the new vector format. I left minor comments and also noticed we need to update `LuceneFilesExtensions` to contain the vector index file `.vex`.

It'd also be good to extend the disk usage analyzer to cover vectors, but this seems best as a follow-up.

dnhatn committed:
minimize changes

dnhatn committed:
Remove ram tree

dnhatn committed:
delay expose analyzers

dnhatn committed:
stylecheck

dnhatn committed:
remove japanese_completion

dnhatn committed:
sort optimization

dnhatn committed:
precommit

dnhatn committed:
fix unknown analyzer

dnhatn reviewed (approved):
I've reviewed all files and I have a question, but changes look great.

romseygeek committed:
Merge remote-tracking branch 'origin/master' into lucene-9-upgrade

romseygeek commented on server/src/main/resources/org/elasticsearch/bootstrap/security.policy:
```diff
@@ -22,6 +22,8 @@ grant codeBase "${codebase.elasticsearch-secure-sm}" {
   // needed for loading plugins which may expect the context class loader to be set
   permission java.lang.RuntimePermission "setContextClassLoader";
+  // needed for SPI class loading
+  permission java.lang.RuntimePermission "accessDeclaredMembers";
```
Not easily, because it's needed by the PluginsService (which used to farm this requirement out to lucene but can't any longer because lucene has removed the class). @rjernst and I have talked very briefly about it, and I see that you've added it to the follow-up list onhttps://github.com/elastic/elasticsearch/issues/74057

romseygeek commented:
> I looked through with a focus on the new vector format. I left minor comments and also noticed we need to update LuceneFilesExtensions to contain the vector index file .vex.

> It'd also be good to extend the disk usage analyzer to cover vectors, but this seems best as a follow-up.

I think both of these come under the disk usage and field usage actions in https://github.com/elastic/elasticsearch/issues/74057

romseygeek committed:
deef
